File: 2020/dec2.py
"""
Solution to https://adventofcode.com/2020/day/2
"""
from collections import Counter
from pathlib import Path

# from the root of the project
INPUT_FILE = Path.cwd() / "2020" / "dec2.txt"


def part1() -> int:
    lines = [line for line in INPUT_FILE.read_text().split("\n") if line]
    valid_passwords = 0
    for line in lines:
        valid_count, letter, password = line.split(" ")
        letter = letter[0]
        low, high = valid_count.split("-")

        c = Counter(password)
        if c[letter] >= int(low) and c[letter] <= int(high):
            valid_passwords += 1

    return valid_passwords


def part2() -> int:
    lines = [line for line in INPUT_FILE.read_text().split("\n") if line]
    valid_passwords = 0
    for line in lines:
        positions, letter, password = line.split(" ")
        letter = letter[0]
        low, high = positions.split("-")

        c_low = password[int(low) - 1] == letter
        c_high = password[int(high) - 1] == letter

        if not (c_low and c_high) and (c_low or c_high):
            valid_passwords += 1

    return valid_passwords


print(part1())
print(part2())

File: 2020/dec2.clj
(ns dec2
  (:require [clojure.string :refer [split-lines]]))

(def input (split-lines (slurp "./dec2.txt")))

(defn is-valid [line]
  (let [[low high c s] (rest ; re-matches returns the whole string as
                             ; well as subgroups, so skip the string
                          (re-matches #"(\d+)-(\d+) (\w): (\w+)" line))
        c-frequency (get (frequencies s) (.charAt c 0) 0)
        low (Integer/parseInt low)
        high (Integer/parseInt high)]
      (<= low c-frequency high)))


(defn part1 []
  (count (filter #(true? (is-valid %)) input)))

(= (part1) 546)

(defn is-valid-2 [line]
  (let [[low high c s] (rest ; re-matches returns the whole string as
                             ; well as subgroups, so skip the string
                          (re-matches #"(\d+)-(\d+) (\w): (\w+)" line))
        c (.charAt c 0)
        v1 (get s (dec (Integer/parseInt low)))
        v2 (get s (dec (Integer/parseInt high)))]
    (or (and (= c v1) (not= c v2))
        (and (= c v2) (not= c v1)))))


(defn part2 []
  (count (filter #(true? (is-valid-2 %)) input)))

(= (part2) 275)
